libs/commons/Textedit.ml

```diff
@@ -53,9 +53,25 @@ let remove_overlapping_edits edits =
   f [] [] edits
 
 let apply_edit_to_text text { start; end_; replacement_text; _ } =
-  let before = Str.first_chars text start in
+  let before = Str.string_before text start in
   let after = Str.string_after text end_ in
-  before ^ replacement_text ^ after
+  let starts_with_newline =
+    String.length before == 0
+    || String.get before (String.length before - 1) == '\n'
+  in
+  let ends_with_newline =
+    String.length after == 0 || String.get after 0 == '\n'
+  in
+  let is_empty_replacement = String.length replacement_text == 0 in
```

Comment:
I agree. I also think that any special business logic should happen outside of the `Textedit` module. A `Textedit.t` is a very simple data structure, and applying it to a file should also be a simple, straightforward operation. If we need to modify the `Textedit.t` to remove newlines, let's do it at the callsite.

Reply:
@nmote It seems difficult to apply the same logic outside of the `apply_edit_to_text` function since we would lose the context of the replacement. Although not impossible, it would require more complex logic to keep track of every change introduced by `TextEdit` module. Instead, we could to create a separate function to remove newlines from empty replacements, which made the code easier to read.

Reply:
I've taken a closer look at how things work today. I still would like to avoid complicating the `apply_edit_to_text` function. Instead, I think that there should be a pass that looks for newlines and modifies the list of `Textedit.t`s to apply, so that they will also replace a newline character as appropriate.

There is already some business logic in `apply_edits`, so I think it would be fine to put that there. And we already have the file contents there, so we wouldn't need to do anything extra to find the additional context.

Also, please make sure to handle CRLF line endings as well.